src/server/service/kv.rs

```diff
@@ -28,18 +31,410 @@ use kvproto::kvrpcpb::{self, *};
+            priority: storage::get_priority_code(ctx.get_priority()),
+            cf,
+        }
+    }
+}
+
+struct ReadBatcher {
+    router: HashMap<ReadId, (Vec<u64>, Vec<PointGetCommand>)>,
+}
+
+impl ReadBatcher {
+    fn new() -> Self {
+        ReadBatcher {
+            router: HashMap::default(),
+        }
+    }
+
+    fn add_get(&mut self, request_id: u64, request: &mut GetRequest) {
+        let id = ReadId::from_context_cf(request.get_context(), None);
+        let command = PointGetCommand::from_get(request);
+        match self.router.get_mut(&id) {
+            Some((reqs, commands)) => {
+                reqs.push(request_id);
+                commands.push(command);
+            }
+            None => {
+                self.router.insert(id, (vec![request_id], vec![command]));
+            }
+        }
+    }
+
+    fn add_raw_get(&mut self, request_id: u64, request: &mut RawGetRequest) {
+        let cf = Some(request.take_cf());
+        let id = ReadId::from_context_cf(request.get_context(), cf);
+        let command = PointGetCommand::from_raw_get(request);
+        match self.router.get_mut(&id) {
+            Some((reqs, commands)) => {
+                reqs.push(request_id);
+                commands.push(command);
+            }
+            None => {
+                self.router.insert(id, (vec![request_id], vec![command]));
+            }
+        }
+    }
+}
+
+impl<E: Engine, L: LockMgr> Batcher<E, L> for ReadBatcher {
+    fn filter(
+        &mut self,
+        request_id: u64,
+        request: &mut batch_commands_request::request::Cmd,
+    ) -> bool {
+        match request {
+            batch_commands_request::request::Cmd::Get(req) => {
+                self.add_get(request_id, req);
+                true
+            }
+            batch_commands_request::request::Cmd::RawGet(req) => {
+                self.add_raw_get(request_id, req);
+                true
+            }
+            _ => false,
+        }
+    }
+
+    fn submit(
+        &mut self,
+        tx: &Sender<(u64, batch_commands_response::Response)>,
+        storage: &Storage<E, L>,
+    ) -> usize {
+        let mut output = 0;
+        for (id, (reqs, commands)) in self.router.drain() {
+            let tx = tx.clone();
+            output += 1;
+            match id.cf {
+                Some(cf) => {
+                    let f = future_raw_batch_get_command(storage, tx, reqs, cf, commands);
+                    poll_future_notify(f);
+                }
+                None => {
+                    let f = future_batch_get_command(storage, tx, reqs, commands);
+                    poll_future_notify(f);
+                }
+            }
+        }
+        output
+    }
+
+    #[inline]
+    fn is_empty(&self) -> bool {
+        self.router.is_empty()
+    }
+}
+
+type ReqBatcherInner<E, L> = (BatchLimiter, Box<dyn Batcher<E, L>>);
+
+struct ReqBatcher<E: Engine, L: LockMgr> {
+    inners: BTreeMap<String, ReqBatcherInner<E, L>>,
```

Comment:
Use an enum of a String should be better?

Reply:
+1